contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,216 @@
+- To verify if user-namespace remapping is enabled at runtime. If found disabled, kubelet will fail to start
+- To determine the default user-namespace range at the runtime, starting UID of which is mapped to the UID '0' of the container.
+
+### Volume Permissions
+Kubelet will change the file permissions at `/var/lib/kubelet/pods` prior to any container start to get file permissions updated according to remapped UID and GID.
+
+### How to disable `NodeUserNamespace` for a specific pod
+This can be done in two ways:
+- **Alpha:** Implicitly using host namespace for the pod containers
+This support is already present (currently it seems broken, will be fixed) in Kubernetes as an experimental functionality, which can be enabled using `feature-gates=”ExperimentalHostUserNamespaceDefaulting=true”`.
+If Pod-Security-Policy is configured to allow the following to be requested by a  pod, host user-namespace will be enabled for the container:
+  - host namespaces (pid, ipc, net)
+  - non-namespaced capabilities (mknod, sys_time, sys_module)
+  - the pod contains a privileged container or using host path volumes.
+  - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L451 introduced via https://github.com/kubernetes/kubernetes/commit/d0d78f478ce0fb9d5e121db3b7c6993b482af82c.
+- **Beta:** Explicit API to request host user-namespace in pod spec
+          This is being targeted under Beta graduation plans.
+
+### CRI API Changes
+Proposed CRI API changes:
+
+```golang
+// Runtime service defines the public APIs for remote container runtimes
+service RuntimeService {
+    // Version returns the runtime name, runtime version, and runtime API version.
+    rpc Version(VersionRequest) returns (VersionResponse) {}
+    …….
+    …….
+    //  RuntimeConfigInfo returns the configuration details of the runtime.
+    rpc RuntimeConfigInfo(RuntimeConfigInfoRequest) returns (RuntimeConfigInfoResponse) {}
+}
+// LinuxIDMapping represents a single user namespace mapping in Linux.
+message LinuxIDMapping {
+   // container_id is the starting id for the mapping inside the container.
+   uint32 container_id = 1;
+   // host_id is the starting id for the mapping on the host.
+   uint32 host_id = 2;
+   // size is the length of the mapping.
+   uint32 size = 3;
+}
+
+// UserNamespaceMapping is a user namespace mapping for uids and gids.
+message UserNamespaceMapping {
+   // uid_mappings is an array of user id mappings.
+   repeated LinuxIDMapping uid_mappings = 1;
+   // gid_mappings is an array of group id mappings.
+   repeated LinuxIDMapping gid_mappings = 2;
+}
+
+message UserNamespaceConfig {
+   // is_enabled, if true indicates that user-namespaces are supported and enabled in the container runtime
+   string is_enabled = 1;
+   // node_usernamespace_mapping is the user namespace mapping set for the container runtime
+   UserNSMapping node_usernamespace_mapping = 2;
```

Comment:
This is a name mismatch to the above `message UserNamespaceMapping`?

But, could we drop this completely and just use `repeated LinuxIDMapping uid_mappings` and `repeated LinuxIDMapping gid_mappings` here directly?

Reply:
+1 on name mismatch. On keeping uid and gids directly in `RuntimeConfig` --  I am trying to follow existing pattern where all the network configurations are abstracted in `NetworkConfig`. If we put uids and gids directly in RuntimeConfig, would not that demand to put all the NetworkConfig members also directly for the sake of consistency?

Reply:
+1